### PR TITLE
apiserver: wire OpenAPI into sample-apiserver

### PR DESCRIPTION
**Pull request by sttts**
This PR add OpenAPI generator to code-generator scripts, and generated and wires the spec in sample-apiserver.

Note: this intentionally checks in the codegen and sample-apiserver spec because the former is a good place to track changes to tooling and the later because an (compilable) example needs it to be checked in. The reason we don't do that at the top-level is because API PRs often conflict at the spec file. This is not the case for codegen and sample-apiserver.

This is also prerequisite for proper separation of the delegated API server specs in https://github.com/kubernetes/kubernetes/pull/74133, which will also clean up technical debt we took when adding CRD openapi publishing, to create the apiextensions-apiserver spec.

```release-note
The sample-apiserver gains support for OpenAPI v2 spec serving at `/openapi/v2`.

The `generate-internal-groups.sh` script in k8s.io/code-generator will generate OpenAPI definitions by default in `pkg/generated/openapi`. Additional API group dependencies can be added via `OPENAPI_EXTRA_PACKAGES=<group>/<version> <group2>/<version2>...`.
```

### CODE REVIEW

**Comment by sttts** [comment]
/assign @cblecker 

For the build/ change.

**Comment by sttts** [comment]
/assign @nikhita 

for the sample-apiserver and code-gen change.

**Review by yue9944882 — APPROVED** [review]
LGTM, i didn't see any openapi marker tags in the example models tho...

**Comment by sttts** [comment]
@yue9944882 because we had them already there. But dysfunctional.

**Comment by cblecker** [comment]
Looks like both the cross build failures and the verify failures are legitimate

File: build/root/.kazelcfg.json [hunk 1]
```diff
@@ -3,7 +3,8 @@
 	"SkippedPaths": [
 		"^_.*",
 		"/_",
-		"^third_party/etcd.*"
+		"^third_party/etcd.*",
+		"k8s.io/sample-apiserver/pkg/apis/.*"
```
Comment (liggitt) [inline build/root/.kazelcfg.json]: no idea what this is doing... surprised we needed this added here if `staging/src/k8s.io/apiextensions-apiserver/pkg/generated/openapi/zz_generated.openapi.go` didn't require it

File: staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh [hunk 2]
```diff
@@ -27,7 +27,7 @@ CODEGEN_PKG=${CODEGEN_PKG:-$(cd "${SCRIPT_ROOT}"; ls -d -1 ./vendor/k8s.io/code-
 #                  instead of the $GOPATH directly. For normal projects this can be dropped.
 CLIENTSET_NAME_VERSIONED=clientset \
 CLIENTSET_NAME_INTERNAL=internalclientset \
-"${CODEGEN_PKG}/generate-internal-groups.sh" all \
+"${CODEGEN_PKG}/generate-internal-groups.sh" deepcopy,client,lister,informer,conversion \
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh]: why remove this?

File: hack/verify-generated-files-remake.sh [hunk 3]
```diff
@@ -291,11 +291,14 @@ fi
 
 assert_clean
 
-touch "staging/src/k8s.io/api/core/v1/types.go"
+touch "staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go"
```
Comment (liggitt) [inline hack/verify-generated-files-remake.sh]: should we touch both files?

File: staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh]: We add openapi to "all", but are not ready for apiextensions-apiserver yet. I have a follow-up to readd it.

File: build/root/.kazelcfg.json (hunk 1 above)
Comment (sttts) [reply build/root/.kazelcfg.json]: Oh, I think we can remove it now that Makefile.generated_files is extended.

File: hack/verify-generated-files-remake.sh (hunk 3 above)
Comment (sttts) [reply hack/verify-generated-files-remake.sh]: Actually I want both tests: apimachinery leads to all specs regenerated, core/v1 only to an updated main spec.

File: build/root/.kazelcfg.json (hunk 1 above)
Comment (sttts) [reply build/root/.kazelcfg.json]: fixed

File: hack/verify-generated-files-remake.sh (hunk 3 above)
Comment (sttts) [reply hack/verify-generated-files-remake.sh]: done

**Comment by sttts** [comment]
@cblecker @liggitt updated and addressed comments. Ptal.

File: staging/src/k8s.io/code-generator/generate-internal-groups.sh [hunk 4]
```diff
@@ -48,6 +48,18 @@ GROUPS_WITH_VERSIONS="$5"
...
+SCRIPT_ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")"; pwd)"
+KUBE_OPENAPI_PKG=$(
+    cd "${SCRIPT_ROOT}"
+    while [ "${PWD}" != "/" ]; do
+        ls -d -1 "${PWD}"/vendor/k8s.io/kube-openapi 2>/dev/null && exit 0
+        cd ..
+    done
+    echo k8s.io/kube-openapi
+)
+go install "${KUBE_OPENAPI_PKG}"/cmd/openapi-gen
```
Comment (liggitt) [inline staging/src/k8s.io/code-generator/generate-internal-groups.sh]: does code-generator have consumers via the published staging repo? what does this addition do for those consumers?

File: staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh]: probably needs to be in the release notes for other consumers of code-gen?

File: hack/verify-generated-files-remake.sh [hunk 5]
```diff
@@ -56,7 +56,7 @@ function newer() {
         if [[ "${F}" -nt "$2" ]]; then
             echo "${F}"
         fi
-    done
+    done | sort
```
Comment (liggitt) [inline hack/verify-generated-files-remake.sh]: some other places add `LC_ALL=C` to avoid locale-specific sort differences... does that matter for these?

**Comment by liggitt** [comment]
a couple minor comments, but otherwise sample-apiserver changes lgtm

I verified the `make generated_files` rebuilds the generated openapi files, but I'm not sure I'm qualified to review the make file changed directly.

File: staging/src/k8s.io/code-generator/generate-internal-groups.sh (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/code-generator/generate-internal-groups.sh]: let me check. I wonder whether we shouldn't put a wrapper into k8s.io/code-generator to reuse the path magic.

File: hack/verify-generated-files-remake.sh (hunk 5 above)
Comment (sttts) [reply hack/verify-generated-files-remake.sh]: those are package names which probably do not fall under locale-specific sorting. But adding LC_ALL=C certainly does not harm.

File: staging/src/k8s.io/code-generator/generate-internal-groups.sh (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/code-generator/generate-internal-groups.sh]: done

File: hack/verify-generated-files-remake.sh (hunk 5 above)
Comment (sttts) [reply hack/verify-generated-files-remake.sh]: done

**Comment by liggitt** [comment]
needs update-vendor.sh to recompute dependencies of codegen

lgtm otherwise. would like ack on the build changes from @cblecker

**Comment by sttts** [comment]
@liggitt updated vendor. @cblecker ptal

File: hack/verify-generated-files-remake.sh [hunk 6]
```diff
@@ -286,12 +286,32 @@ if [[ -n "${X}" ]]; then
...
+fi
+
+#
+# Test when we touch a file in a package that needs codegen for only the main openapi spec.
 #
 
 assert_clean
 
-touch "staging/src/k8s.io/api/core/v1/types.go"
+touch "staging/src/k8s.io/api/apps/v1/types.go"
```
Comment (spiffxp) [inline hack/verify-generated-files-remake.sh]: why the change from core to apps here?

File: staging/src/k8s.io/code-generator/generate-internal-groups.sh [hunk 7]
```diff
@@ -108,3 +109,14 @@ if [ "${GENS}" = "all" ] || grep -qw "informer" <<<"${GENS}"; then
            --output-package "${OUTPUT_PKG}/informers" \
            "$@"
 fi
+
+if [ "${GENS}" = "all" ] || grep -qw "openapi" <<<"${GENS}"; then
+  echo "Generating OpenAPI definitions for ${GROUPS_WITH_VERSIONS} at ${OUTPUT_PKG}/openapi"
+  declare -a OPENAPI_EXTRA_PACKAGES
```
Comment (spiffxp) [inline staging/src/k8s.io/code-generator/generate-internal-groups.sh]: why is this needed? we don't appear to be populating it at all

(also arrays in bash is a sign to start considering something other than bash)

**Review by spiffxp — COMMENTED** [review]
Didn't have time to wade through the Makefile, but some questions about the bash surrounding it

File: staging/src/k8s.io/code-generator/generate-internal-groups.sh (hunk 7 above)
Comment (sttts) [reply staging/src/k8s.io/code-generator/generate-internal-groups.sh]: because it will be needed for apiextensions-apiserver (follow-up PR) which needs autoscaling/v1 as additional group.

File: hack/verify-generated-files-remake.sh (hunk 6 above)
Comment (sttts) [reply hack/verify-generated-files-remake.sh]: Because core/v1 is a dependency for autoscaling/v1, which is needed by apiextensions-apiserver (in a follow-up). apps/v1 is only needed by kube-apiserver, and therefore is a better choice here.

File: staging/src/k8s.io/code-generator/generate-internal-groups.sh (hunk 7 above)
Comment (sttts) [reply staging/src/k8s.io/code-generator/generate-internal-groups.sh]: > also arrays in bash is a sign to start considering something other than bash

It's our choice in k8s.io/code-generator for project not using kubebuilder.

The `declare -a` makes sure that `${OPENAPI_EXTRA_PACKAGES[@]}` in the next line does not fail when the variable is empty.

File: staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh]: added

**Review by fejta — COMMENTED** [review]
/approve

File: staging/src/k8s.io/code-generator/generate-internal-groups.sh (hunk 7 above)
Comment (spiffxp) [reply staging/src/k8s.io/code-generator/generate-internal-groups.sh]: Yeah, I understand this is the language you're locked into, and it's a sane/clean declaration given that constraint. I just wanted to make sure this wasn't something being left around just incase, follow on makes sense.

**Comment by spiffxp** [comment]
/lgtm
That Makefile is pretty eye-watering stuff, but the changes are sanely written for what it is.  Thanks for the consistent ordering and use of `prefix` and the answers to my bash questions.  I can't speak to the dependencies and such being correct, but I'll trust that to the /approve

**Comment by cblecker** [comment]
Reviewed and mostly looks sane. One open question, otherwise LGTM.

File: build/root/Makefile.generated_files [hunk 8]
```diff
@@ -408,81 +408,97 @@ $(CONVERSION_GEN): $(k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd/conversi
...
 BOILERPLATE_FILENAME := vendor/k8s.io/code-generator/hack/boilerplate.go.txt
 REPORT_FILENAME := $(OUT_DIR)/violations.report
 IGNORED_REPORT_FILENAME := $(OUT_DIR)/ignored_violations.report
-KNOWN_VIOLATION_FILENAME := api/api-rules/violation_exceptions.list
+
 # When UPDATE_API_KNOWN_VIOLATIONS is set to be true, let the generator to write
 # updated API violations to the known API violation exceptions list.
 ifeq ($(UPDATE_API_KNOWN_VIOLATIONS),true)
-    REPORT_FILENAME:=$(KNOWN_VIOLATION_FILENAME)
+    REPORT_FILENAME:=$(KUBE_KNOWN_VIOLATION_FILENAME)
```
[the remark this answers is not in the document]
Comment (sttts) [reply build/root/Makefile.generated_files]: Fixed.

This code path was even broken before this PR because we forgot the dependency from the violation file (which is touched) to the output. I fixed that. And I added a test case.

**Comment by sttts** [comment]
Rebased on changed ObjectMeta comment.
